### ISSUE TITLE
lineinfile module behavior

### ISSUE THREAD

**Issue by sterfield** [ansible/ansible-modules-core#1884]
Hi,

I want to have a discussion about "lineinfile" module specifications, and the behavior in several cases.

Here's the algorithm of lineinfile module, when "state=present" :
### line mentioned

Searching for "line".
- If line is found, then do nothing
- If line is not found, then add the line at the very end of the file (because if insertafter and insertbefore are not filled, then insertafter is forced to EOF).
### line mentioned and insertbefore=BOF
- If line is found, then do nothing
- If line is not found, then add the line at the beginning of the file
### line mentioned and insertbefore=BOF and regexp  = regexp2
- if line is found then do nothing
- if line is not found and regexp is found then change the "regexp2" line by "line" (insertbefore is ignored)
- if line is not found and regexp is not found, then line is added at the beginning of the file.
### line mentioned and insertbefore = regexp
- if line is found, then do nothing
- if line is not found, and insertbefore regexp is not found, then add the line at the end of the file
- If line is not found, and insertbefore regexp is found, then add the line before the 'insertbefore' line.
### line mentioned and insertbefore = regexp and regexp = regexp2
- if line is found, then do nothing
- if line is not found and insertbefore regexp is found and regexp is found then change the "regepx2" line by "line"
- if line is not found and insertbefore regexp is found and regexp is not found, then insert "line" before "regexp" line
- if line is not found and insertbefore regexp is not found and regexp is found, then change "regexp2" by "line"
- if line is not found and insertbefore regexp is not found and regexp is not found, then add the line at the end of the file.

It's the same for insertafter.

Now, considering the behavior describes here-before, I have several comments :
- I can understand that, if you search for a line without mentioning "regexp", and you don't find the line in the file, you add it at the end of the file. The description of the module is pretty explicit "This module will search a file for a line, and ensure that it is present or absent."
- However, IMHO, when you specify a regexp, it means that you want to change a line. So, to me, if the regexp is not matched, the file should be left intact, without any addition / modification. The documentation is not clear about the behavior of the module in this case.

So I would like to discuss with you guys, the behavior of the module when you use "regexp" (and possibly insertbefore, insertafter). I would like also to hear the thoughts of the some Ansible users that think it's a normal behavior to have regexp mentioned, that doesn't matched, but have the line still added at the end of the file.

Thanks,

**Comment by bcoca** [comment]
# List Information

Hi!

Thanks very much for your interest in Ansible.  It sincerely means a lot to us. 

This appears to be a user question, and we'd like to direct these kinds of things to either the mailing list or the IRC channel.
- IRC: #ansible on irc.freenode.net  
- mailing list: https://groups.google.com/forum/#!forum/ansible-project
- mailing list: https://groups.google.com/forum/#!forum/ansible-devel

If you can stop by there, we'd appreciate it.  This allows us to keep the issue tracker for bugs, pull requests, RFEs and the like.

Thank you once again and we look forward to seeing you on the list or IRC.  Thanks!

**Comment by sterfield** [comment]
Hi Brian,

I did opened a thread on the mailing list. Here's the link : https://groups.google.com/forum/#!topic/ansible-project/riuUuT7LJ0I

But quite frankly, I did not have any answer nor comments from anyone, that's why I've opened an issue here.

Honestly, the fact that "lineinfile" is putting the "line" at the end of the file when the "regexp" is not matched sounds like a bug to me. But I wanted to talk about it first, because I may miss something here.

I'm also OK to provide a PR on this issue, but first, I think that we should describe the "specifications" of the "lineinfile" module.

Hence the Github issue.

Thanks for your help,
